### PR TITLE
Rename Pulsar Beamformer to Tracking Beamformer

### PR DESCRIPTION
**Pull request by james-s-willis**

### CODE REVIEW

File: config/chime_science_run_gpu.yaml [hunk 1]
```diff
@@ -439,22 +439,22 @@ vis_buffers:
   visbuf_eig_10s_merge:
     kotekan_buffer: vis
   # Buffers for gated
-  visbuf_psr0_5s_0:
+  visbuf_trk0_5s_0:
```
Comment (andrerenard) [inline config/chime_science_run_gpu.yaml:442]: I think these are unrelated buffers. In this case `psr` is talking about gated N^2 pulsar data...

File: config/chime_psr_verification.yaml [hunk 2]
```diff
@@ -1,8 +1,8 @@
 ##########################################
 #
-# PSR verification
+# TRACKING verification
```
Comment (andrerenard) [inline config/chime_psr_verification.yaml:3]:
```suggestion
# Tracking Beamformer verification
```

File: config/chime_psr_verification.yaml [hunk 3]
```diff
@@ -1,8 +1,8 @@
...
 #
-# Runs both GPU and CPU PSR beamformer pipelines, include re-ordering, read gains files, calculate phase shift, brute-force beamforming
+# Runs both GPU and CPU TRACKING beamformer pipelines, include re-ordering, read gains files, calculate phase shift, brute-force beamforming
```
Comment (andrerenard) [inline config/chime_psr_verification.yaml:5]:
```suggestion
# Runs both GPU and CPU tracking beamformer pipelines, include re-ordering, read gains files, calculate phase shift, brute-force beamforming
```

File: config/chime_psr_verification.yaml [hunk 4]
```diff
@@ -245,18 +245,18 @@ gpu:
     kotekan_stage: hsaProcess
     gpu_id: 0
     updatable_config:
-      gain_psr: /updatable_config/pulsar_gain
-      psr_pt: /updatable_config/pulsar_pointing
+      gain_tracking: /updatable_config/pulsar_gain
+      tracking_pt: /updatable_config/pulsar_pointing
```
Comment (andrerenard) [inline config/chime_psr_verification.yaml:249]: We should also update the names for the endpoints since we do that in the other configs.

File: config/chime_science_run_gpu.yaml [hunk 5]
```diff
@@ -211,11 +211,11 @@ updatable_config:
     # not correlator/FPGA input order.
     element_index: 500
   gating:
-    psr0_config:
+    tracking0_config:
       kotekan_update_endpoint: "json"
       enabled: false
       # B0329 (2018/11/14)
-      pulsar_name: "B0329"
+      tracking_name: "B0329"
```
Comment (andrerenard) [inline config/chime_science_run_gpu.yaml:218]: This is a part of the N^2 pulsar gating engine.  It shouldn't be changed.

File: config/chime_science_run_gpu.yaml [hunk 6]
```diff
@@ -1420,37 +1420,37 @@ vis_accumulate:
...
         buf: visbuf_psr0_5s_2
   acc3:
     kotekan_stage: visAccumulate
     in_buf: rfi_vis_buffer_3
     out_buf: visbuf_5s_3
     gating:
-      psr0:
+      tracking0:
         mode: pulsar
         buf: visbuf_psr0_5s_3
```
Comment (andrerenard) [inline config/chime_science_run_gpu.yaml:1455]: These are all part of the N^2 pulsar gating system, and shouldn't be updated.

File: config/chime_science_run_gpu.yaml [hunk 7]
```diff
@@ -1562,7 +1562,7 @@ buffer_send:
     kotekan_stage: bufferSend
     buf: visbuf_5s_26m
     server_port: 11025
-  26m_psr0:
+  26m_tracking0:
```
Comment (andrerenard) [inline config/chime_science_run_gpu.yaml:1565]: Same here

File: config/chime_science_run_gpu.yaml [hunk 8]
```diff
@@ -1465,7 +1465,7 @@ vis_merge_5s:
     - visbuf_5s_3
   out_buf: visbuf_5s_merge
 
-vis_merge_gated_psr0:
+vis_merge_gated_tracking0:
```
Comment (andrerenard) [inline config/chime_science_run_gpu.yaml:1468]: Same here

File: config/chime_science_run_gpu.yaml [hunk 9]
```diff
@@ -1542,7 +1542,7 @@ vis_debug:
   input_list: [1225, 1521]  # 26m channels
 
 # Generate the 26m gated stream
-26m_psr0_subset:
+26m_tracking0_subset:
```
Comment (andrerenard) [inline config/chime_science_run_gpu.yaml:1545]: Same here

File: config/chime_science_run_gpu_no_output.yaml [hunk 10]
```diff
@@ -210,11 +210,11 @@ updatable_config:
     # not correlator/FPGA input order.
     element_index: 500
   gating:
-    psr0_config:
+    tracking0_config:
       kotekan_update_endpoint: "json"
       enabled: false
       # B0329 (2018/11/14)
-      pulsar_name: "B0329"
+      tracking_name: "B0329"
```
Comment (andrerenard) [inline config/chime_science_run_gpu_no_output.yaml:217]: Same comments apply to all the N^2 gating related changes in this file as well.

File: config/frb_test.yaml [hunk 11]
```diff
@@ -1,6 +1,6 @@
 ##########################################
 #
-# FRB / PSR verification
+# FRB / TRACKING verification
```
Comment (andrerenard) [inline config/frb_test.yaml:3]:
```suggestion
# FRB / Tracking Beamformer verification
```

File: lib/hsa/hsaTrackingBeamform.hpp [hunk 12]
```diff
@@ -58,32 +60,33 @@
...
 
     /// Allocate kernel argument buffer, set kernel dimensions, enqueue kernel
     hsa_signal_t execute(int gpu_frame_id, hsa_signal_t precede_signal) override;
 
 private:
     /// Input length, should be nsamp x n_elem x 2 for complex / 2 since we pack two 4-bit in one
     int32_t input_frame_len;
-    /// Output length, should be 10psr x nsamp x 2 pol x 2 for complex / 2 since we pack two 4-bit
-    /// in one
+    /// Output length, should be 10tracking x nsamp x 2 pol x 2 for complex / 2 since we pack two
```
Comment (andrerenard) [inline lib/hsa/hsaTrackingBeamform.hpp:79]:
```suggestion
    /// Output length, should be num_beams x nsamp x 2 pol x 2 for complex / 2 since we pack two
```
I think that should be `num_beams` since this is no longer fixed at 10

File: lib/hsa/hsaTrackingUpdatePhase.hpp [hunk 13]
```diff
@@ -49,43 +49,43 @@
...
  * @conf   num_elements         Int (default 2048). Number of elements
- * @conf   num_beams            Int (default 10). Number of pulsars
+ * @conf   num_beams            Int (default 10). Number of beams
  * @conf   feed_sep_NS          Float (default 0.3048). N-S feed separation in m.
  * @conf   feed_sep_EW          Float (default 22.0). E-W feed separation in m.
  * missing
- * @conf   pulsar_pointing/i/ra       Float - initial RA (in deg) to form beams on for beam_id=i
- * @conf   pulsar_pointing/i/dec      Float - initial Dec (in deg) to form beams on for beam_id=i
- * @conf   pulsar_pointing/i/scaling  Int - nominal scaling for beam_id=i
+ * @conf   beams_pointing/i/ra       Float - initial RA (in deg) to form beams on for beam_id=i
```
Comment (andrerenard) [inline lib/hsa/hsaTrackingUpdatePhase.hpp:56]:
```suggestion
 * @conf   beam_pointing/i/ra       Float - initial RA (in deg) to form beams on for beam_id=i
```

File: lib/hsa/hsaTrackingUpdatePhase.hpp [hunk 14]
```diff
@@ -141,16 +141,16 @@ class hsaPulsarUpdatePhase : public hsaCommand {
...
 
     /// Flag to avoid re-calculating freq-specific params except at first pass
     bool first_pass;
     /// Flag to trigger phase update, either because of endpt message or every second
     bool update_phase;
 
-    /// Endpoint for updating psr coordinates
-    std::string endpoint_psrcoord;
+    /// Endpoint for updating tracking coordinates
+    std::string endpoint_trackingcoord;
```
Comment (andrerenard) [inline lib/hsa/hsaTrackingUpdatePhase.hpp:153]:
```suggestion
    std::string endpoint_beam_coord;
```
This is maybe more consistent with the other names above.

File: config/chime_science_run_gpu.yaml [hunk 15]
```diff
@@ -103,9 +103,9 @@ updatable_config:
   frb_gain:
     kotekan_update_endpoint: json
     frb_gain_dir: /mnt/frb-archiver/daily_gain_solutions/Latest_FRB
-  pulsar_gain:
+  tracking_gain:
```
Comment (andrerenard) [inline config/chime_science_run_gpu.yaml:106]: This should have a parallel PR in `coco` to update the names of these endpoints.  And the Pulsar group will also need to update their scripts.

File: lib/hsa/kernels/tracking_beamformer.it.cl [hunk 16]
```diff
@@ -29,7 +29,7 @@
 #define TS_BUNDLE       TS_RED
 #define BM_BUNDLE       BM_RED
 
- __kernel void pulsarbf(
+ __kernel void trackingbf(
```
Comment (andrerenard) [inline lib/hsa/kernels/tracking_beamformer.it.cl:32]: Can you just delete the beamforming kernels we aren't using?

File: lib/stages/ReadGain.cpp [hunk 17]
```diff
@@ -200,44 +201,44 @@ void ReadGain::read_gain_psr() {
...
-    mark_frame_full(gain_psr_buf, unique_name.c_str(), gain_psr_buf_id);
-    DEBUG("Maked gain_psr_buf frame {:d} full", gain_psr_buf_id);
-    INFO("Time required to load PSR gains: {:f}", current_time() - start_time);
-    DEBUG("Gain_psr_buf: {:.2f} {:.2f} {:.2f} ", out_frame_psr[0], out_frame_psr[1],
-          out_frame_psr[2]);
-    gain_psr_buf_id = (gain_psr_buf_id + 1) % gain_psr_buf->num_frames;
+    gains_last_update_timestamp_metric.labels({"tracking"}).set(start_time);
+    mark_frame_full(gain_tracking_buf, unique_name.c_str(), gain_tracking_buf_id);
+    DEBUG("Maked gain_tracking_buf frame {:d} full", gain_tracking_buf_id);
+    INFO("Time required to load TRACKING gains: {:f}", current_time() - start_time);
```
Comment (andrerenard) [inline lib/stages/ReadGain.cpp:238]:
```suggestion
    INFO("Time required to load tracking beamformer gains: {:f}", current_time() - start_time);
```

File: lib/stages/ReadGain.hpp [hunk 18]
```diff
@@ -1,6 +1,6 @@
 /**
  * @file
- * @brief read in new gain file for FRB/PSR when available
+ * @brief read in new gain file for FRB/TRACKING when available
```
Comment (andrerenard) [inline lib/stages/ReadGain.hpp:3]:
```suggestion
 * @brief read in new gain file for FRB/Tracking beamformer when available
```

File: lib/stages/ReadGain.hpp [hunk 19]
```diff
@@ -26,13 +26,13 @@ using std::vector;
 
 /**
  * @class ReadGain
- * @brief read in new gain files for FRB/PSR when available
+ * @brief read in new gain files for FRB/TRACKING when available
```
Comment (andrerenard) [inline lib/stages/ReadGain.hpp:29]:
```suggestion
 * @brief read in new gain files for FRB/Tracking beamformer when available
```

File: lib/stages/ReadGain.hpp [hunk 20]
```diff
@@ -52,7 +52,7 @@ using std::vector;
  *
  * The gain path is registered as a subscriber to an updatable config block.
  * For the FRB, it is one directory path: '{"frb_gain_dir":"the_new_path"}'
- * For the PSR, it is an array of 10 paths for each of the 10 beams:
+ * For the TRACKING, it is an array of 10 paths for each of the 10 beams:
```
Comment (andrerenard) [inline lib/stages/ReadGain.hpp:55]:
```suggestion
 * For the tracking beamformer, it is an array of @c num_beams paths for each of the @c num_beams beams:
```

File: lib/stages/ReadGain.hpp [hunk 21]
```diff
@@ -69,13 +69,13 @@ class ReadGain : public kotekan::Stage {
     /// Endpoint for providing new directory path for FRB gain updates
     bool update_gains_frb_callback(nlohmann::json& json);
 
-    /// Endpoint for providing new directory path for PSR gain updates
-    bool update_gains_psr_callback(nlohmann::json& json);
+    /// Endpoint for providing new directory path for TRACKING gain updates
```
Comment (andrerenard) [inline lib/stages/ReadGain.hpp:72]:
```suggestion
    /// Endpoint for providing new directory path for tracking beamformer gain updates
```

File: lib/stages/ReadGain.hpp [hunk 22]
```diff
@@ -69,13 +69,13 @@ class ReadGain : public kotekan::Stage {
...
+    bool update_gains_tracking_callback(nlohmann::json& json);
 
     /// Read gain file for frb
     void read_gain_frb();
-    /// Read gain file for psr
-    void read_gain_psr();
+    /// Read gain file for tracking
```
Comment (andrerenard) [inline lib/stages/ReadGain.hpp:77]:
```suggestion
    /// Read gain file for tracking beamformer
```

File: lib/testing/gpuTrackingBeamformSimulate.hpp [hunk 23]
```diff
@@ -56,8 +56,8 @@ class gpuBeamformPulsarSimulate : public kotekan::Stage {
     uint32_t _num_elements;
     /// number of samples = 49152
     int32_t _samples_per_data_set;
-    /// number of pulsars = 10
-    int32_t _num_pulsar;
+    /// number of beams = 10
```
Comment (andrerenard) [inline lib/testing/gpuTrackingBeamformSimulate.hpp:59]:
```suggestion
    /// number of beams
```

File: lib/testing/gpuTrackingBeamformSimulate.hpp [hunk 24]
```diff
@@ -94,8 +94,8 @@ class gpuBeamformPulsarSimulate : public kotekan::Stage {
     /// Time now in second
     struct timespec time_now_gps;
 
-    /// 10 pulsar RA, DEC and scaling factor
-    struct psrCoord psr_coord; // active coordinates to be passed to metatdata
+    /// 10 beams RA, DEC and scaling factor
```
Comment (andrerenard) [inline lib/testing/gpuTrackingBeamformSimulate.hpp:97]:
```suggestion
    /// RA, DEC and scaling factor for the tracking beams
```

File: lib/testing/gpuTrackingBeamformSimulate.hpp [hunk 25]
```diff
@@ -105,14 +105,14 @@ class gpuBeamformPulsarSimulate : public kotekan::Stage {
     bool update_phase;
 
     /// Brute-force beamform by multiplying each element with a phase shift
-    void cpu_beamform_pulsar(double* input, double* phase, float* output, int nsamp, int nelem,
-                             int npsr, int npol);
+    void cpu_beamform_tracking(double* input, double* phase, float* output, int nsamp, int nelem,
```
Comment (andrerenard) [inline lib/testing/gpuTrackingBeamformSimulate.hpp:108]:
```suggestion
    void cpu_tracking_beamformer(double* input, double* phase, float* output, int nsamp, int nelem,
```
In other places we used "tracking beamformer", we should be consistent with the order of names if possible.

File: config/chime_science_run_gpu.yaml (hunk 15 above)
Comment (james-s-willis) [reply config/chime_science_run_gpu.yaml:106]: I've created https://github.com/chime-experiment/coco/pull/216 for this change in `coco`

File: config/tests/chime_frb_verification.yaml [hunk 26]
```diff
@@ -1,6 +1,6 @@
 ##########################################
 #
-# FRB / PSR verification
+# FRB / TRACKING verification
```
Comment (andrerenard) [inline config/tests/chime_frb_verification.yaml:3]:
```suggestion
# FRB / Tracking beamformer verification
```

File: lib/hsa/kernels/tracking_beamformer_nbeam.cl [hunk 27]
```diff
@@ -3,7 +3,7 @@
 #define IM y
 #define NBEAM 1
```
Comment (andrerenard) [inline lib/hsa/kernels/tracking_beamformer_nbeam.cl:4]: Just because I'd like to understand, do you understand what does this variable does?  It looks like `num_beams` is set by one of the `global_id`, so then what does `NBEAM` do?  My guess is it controls the number of beams done within the kernel, but honestly it's a bit confusing with `num_beams` being another variable going into this.

File: lib/hsa/hsaTrackingBeamform.hpp [hunk 28]
```diff
@@ -45,11 +47,11 @@
...
  *
  * @conf   num_elements         Int (default 2048). Number of elements
- * @conf   num_pulsar           Int (default 10). Number of pulsar beams to be formed
+ * @conf   num_beams           Int (default 10). Number of beams to be formed
  * @conf   samples_per_data_set Int (default 49152). Number of time samples in a data set
  * @conf   num_pol              Int (default 2). Number of polarizations
- * @conf   command              String (defualt: "pulsarbf"). Kernel command.
- * @conf   kernel               String (default: "pulsar_beamformer.hsaco"). Kernel filename.
+ * @conf   command              String (defualt: "trackingbf"). Kernel command.
+ * @conf   kernel               String (default: "tracking_beamformer.hsaco"). Kernel filename.
```
Comment (andrerenard) [inline lib/hsa/hsaTrackingBeamform.hpp:54]:
```suggestion
 * @conf   kernel               String (default: "tracking_beamformer_nbeam.hsaco"). Kernel filename.
```

File: lib/hsa/hsaTrackingBeamform.hpp [hunk 29]
```diff
@@ -45,11 +47,11 @@
  *     @gpu_mem_metadata     none
  *
  * @conf   num_elements         Int (default 2048). Number of elements
- * @conf   num_pulsar           Int (default 10). Number of pulsar beams to be formed
+ * @conf   num_beams           Int (default 10). Number of beams to be formed
  * @conf   samples_per_data_set Int (default 49152). Number of time samples in a data set
  * @conf   num_pol              Int (default 2). Number of polarizations
- * @conf   command              String (defualt: "pulsarbf"). Kernel command.
- * @conf   kernel               String (default: "pulsar_beamformer.hsaco"). Kernel filename.
+ * @conf   command              String (defualt: "trackingbf"). Kernel command.
```
Comment (andrerenard) [inline lib/hsa/hsaTrackingBeamform.hpp:53]:
```suggestion
 * @conf   command              String (defualt: "trackingbf_float"). Kernel command.
```

File: lib/stages/ReadGain.cpp [hunk 30]
```diff
@@ -114,26 +115,26 @@ bool ReadGain::update_gains_frb_callback(nlohmann::json& json) {
...
+        std::string output_msg = "[TRACKING] Updating gains from ";
         for (int i = 0; i < _num_beams; i++) {
-            output_msg += _gain_dir_psr[i];
+            output_msg += _gain_dir_tracking[i];
             output_msg += " ";
         }
         INFO("{:s}", output_msg);
     } catch (std::exception const& e) {
-        WARN("[PSR] Fail to read gain_dir {:s}", e.what());
+        WARN("[TRACKING] Fail to read gain_dir {:s}", e.what());
```
Comment (andrerenard) [inline lib/stages/ReadGain.cpp:132]:
```suggestion
        WARN("[Tracking Beamformer] Fail to read gain_dir {:s}", e.what());
```

**Comment by andrerenard** [comment]
Oh and you'll also need to make changes to `ch_ansible`: https://github.com/chime-experiment/ch_ansible/tree/master/roles/internal/coco_common/files/endpoints

File: lib/hsa/kernels/tracking_beamformer_nbeam.cl (hunk 27 above)
Comment (james-s-willis) [reply lib/hsa/kernels/tracking_beamformer_nbeam.cl:4]: Yeah I'm not sure why it's there either. It doesn't do anything:
```
for (int b=get_global_id(1)*NBEAM; b<(get_global_id(1)+1)*NBEAM; b++){
```
should I just remove it?

Comment (andrerenard) [reply lib/hsa/kernels/tracking_beamformer_nbeam.cl:4]: Hmm, let's leave it for now.   I think it might be an optimization parameter.  I'll check with Keith about that kernel, since I think he was the last one to edit it.

**Comment by andrerenard** [comment]
Looks like it failed clang-format

File: lib/hsa/kernels/tracking_beamformer_nbeam.cl (hunk 27 above)
Comment (james-s-willis) [reply lib/hsa/kernels/tracking_beamformer_nbeam.cl:4]: Okay

**Comment by james-s-willis** [comment]
I've created a PR in `ch_ansible`: https://github.com/chime-experiment/ch_ansible/pull/124

File: lib/hsa/hsaTrackingBeamform.hpp [hunk 31]
```diff
@@ -16,19 +16,21 @@
...
- * The default number of pulsar beams to be formed is 10. The phases are matrix
+ * This is an hsaCommand that launches the kernel (tracking_beamformer) for
+ * brute-force coherent beamforming and is most applicable to tracking observations.
+ * An array of phases (shape @c n_tracking x @c n_elem x 2) is calculated by
+ hsaTrackingUpdatePhase.cpp.
+ * The default number of tracking beams to be formed is 10. The phases are matrix
  * multiplied with the input data (shape @c n_samp x @c n_elem) and the output is of dimension
- * (@c n_samp x @c n_psr x @c n_pol x 2). Output data type is 4-4b int packed as char. Currently
+ * (@c n_samp x @c n_tracking x @c n_pol x 2). Output data type is 4-4b int packed as char.
+ Currently
```
Comment (andrerenard) [inline lib/hsa/hsaTrackingBeamform.hpp:30]: I'm not sure why, but the clang-format isn't adding an extra `* ` to the start of these lines.   Same on line `55` where it would be nice to also line up the new line with sentence above.

File: config/chime_science_run_gpu.yaml (hunk 15 above)
Comment (cubranic) [reply config/chime_science_run_gpu.yaml:106]: And the Pulsar script is changed in CHIME-Pulsar-Timing/ch_psr_gains#7.
